The following notebook is a movie genre classification and it is a part of codsoft internship. In the below notebook I used CounterVectorizer to convert text into vectors and further trained models like Logistic Regression,Random Forest Classifier and KNN Classifier.




###EDA & Text Processing :

In [ ]:
#Importing Basic Libraries
import numpy as np
import pandas as pd
import seaborn as sns

In [ ]:
#Reading the Data
trainDf = pd.read_table("train_data.txt", delimiter=":::", names=["id", "title", "genre", "description"])
testDf=pd.read_table('test_data.txt',delimiter=":::",names=['id','title','description'])

<ipython-input-6-797e9dd35d19>:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  trainDf = pd.read_table("train_data.txt", delimiter=":::", names=["id", "title", "genre", "description"])
<ipython-input-6-797e9dd35d19>:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  testDf=pd.read_table('test_data.txt',delimiter=":::",names=['id','title','description'])


In [ ]:
trainDf.head()

id                               title       genre  \
0   1       Oscar et la dame rose (2009)       drama    
1   2                       Cupid (1997)    thriller    
2   3   Young, Wild and Wonderful (1980)       adult    
3   4              The Secret Sin (1915)       drama    
4   5             The Unrecovered (2007)       drama    

                                         description  
0   Listening in to a conversation between his do...  
1   A brother and sister with a past incestuous r...  
2   As the bus empties the students for their fie...  
3   To help their unemployed father make ends mee...  
4   The film's title refers not only to the un-re...

In [ ]:
testDf.head()

id                          title  \
0   1          Edgar's Lunch (1998)    
1   2      La guerra de papá (1977)    
2   3   Off the Beaten Track (2010)    
3   4        Meu Amigo Hindu (2015)    
4   5             Er nu zhai (1955)    

                                         description  
0   L.R. Brane loves his life - his car, his apar...  
1   Spain, March 1964: Quico is a very naughty ch...  
2   One year in the life of Albin and his family ...  
3   His father has died, he hasn't spoken with hi...  
4   Before he was known internationally as a mart...

In [ ]:
trainDf.isna().sum() #Checking The Null Values

id             0
title          0
genre          0
description    0
dtype: int64

In [ ]:
testDf.isna().sum()

id             0
title          0
description    0
dtype: int64

Merging title and description.

In [ ]:
trainDf['Movie']=trainDf['title']+trainDf['description'] #Merging Title & Description to create 'Movie' columns

In [ ]:
trainDf.head()

id                               title       genre  \
0   1       Oscar et la dame rose (2009)       drama    
1   2                       Cupid (1997)    thriller    
2   3   Young, Wild and Wonderful (1980)       adult    
3   4              The Secret Sin (1915)       drama    
4   5             The Unrecovered (2007)       drama    

                                         description  \
0   Listening in to a conversation between his do...   
1   A brother and sister with a past incestuous r...   
2   As the bus empties the students for their fie...   
3   To help their unemployed father make ends mee...   
4   The film's title refers not only to the un-re...   

                                               Movie  
0   Oscar et la dame rose (2009)  Listening in to...  
1   Cupid (1997)  A brother and sister with a pas...  
2   Young, Wild and Wonderful (1980)  As the bus ...  
3   The Secret Sin (1915)  To help their unemploy...  
4   The Unrecovered (2007)  The film's title refe...

In [ ]:
trainDf.value_counts('genre') #Knowing the value counts of each genre

genre
 drama           13613
 documentary     13096
 comedy           7447
 short            5073
 horror           2204
 thriller         1591
 action           1315
 western          1032
 reality-tv        884
 family            784
 adventure         775
 music             731
 romance           672
 sci-fi            647
 adult             590
 crime             505
 animation         498
 sport             432
 talk-show         391
 fantasy           323
 mystery           319
 musical           277
 biography         265
 history           243
 game-show         194
 news              181
 war               132
dtype: int64

In [ ]:
testDf['Movie']=testDf['title']+testDf['description'] #Merging title & description

In [ ]:
testDf.head()

id                          title  \
0   1          Edgar's Lunch (1998)    
1   2      La guerra de papá (1977)    
2   3   Off the Beaten Track (2010)    
3   4        Meu Amigo Hindu (2015)    
4   5             Er nu zhai (1955)    

                                         description  \
0   L.R. Brane loves his life - his car, his apar...   
1   Spain, March 1964: Quico is a very naughty ch...   
2   One year in the life of Albin and his family ...   
3   His father has died, he hasn't spoken with hi...   
4   Before he was known internationally as a mart...   

                                               Movie  
0   Edgar's Lunch (1998)  L.R. Brane loves his li...  
1   La guerra de papá (1977)  Spain, March 1964: ...  
2   Off the Beaten Track (2010)  One year in the ...  
3   Meu Amigo Hindu (2015)  His father has died, ...  
4   Er nu zhai (1955)  Before he was known intern...

Dropping id,title and description.

In [ ]:
trainDf.drop(columns=['id',"title", "description"], axis=1, inplace=True) #Dropping the columns which are not required

In [ ]:
trainDf.head()

genre                                              Movie
0      drama    Oscar et la dame rose (2009)  Listening in to...
1   thriller    Cupid (1997)  A brother and sister with a pas...
2      adult    Young, Wild and Wonderful (1980)  As the bus ...
3      drama    The Secret Sin (1915)  To help their unemploy...
4      drama    The Unrecovered (2007)  The film's title refe...

In [ ]:
testDf.drop(columns=['id','title','description'],axis=1,inplace=True)

In [ ]:
testDf.head()

Movie
0   Edgar's Lunch (1998)  L.R. Brane loves his li...
1   La guerra de papá (1977)  Spain, March 1964: ...
2   Off the Beaten Track (2010)  One year in the ...
3   Meu Amigo Hindu (2015)  His father has died, ...
4   Er nu zhai (1955)  Before he was known intern...

Converting 'Genre' into number becasue machine learning models doesn't understand text.

In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
encoder = LabelEncoder()
trainDf["genre"] = encoder.fit_transform(trainDf["genre"])

In [ ]:
trainDf.head()

genre                                              Movie
0      8   Oscar et la dame rose (2009)  Listening in to...
1     24   Cupid (1997)  A brother and sister with a pas...
2      1   Young, Wild and Wonderful (1980)  As the bus ...
3      8   The Secret Sin (1915)  To help their unemploy...
4      8   The Unrecovered (2007)  The film's title refe...

Identifying input and target

In [ ]:
train_input=trainDf['Movie']

In [ ]:
train_target=trainDf['genre']

In [ ]:
train_input

0         Oscar et la dame rose (2009)  Listening in to...
1         Cupid (1997)  A brother and sister with a pas...
2         Young, Wild and Wonderful (1980)  As the bus ...
3         The Secret Sin (1915)  To help their unemploy...
4         The Unrecovered (2007)  The film's title refe...
                               ...                        
54209     "Bonino" (1953)  This short-lived NBC live si...
54210     Dead Girls Don't Cry (????)  The NEXT Generat...
54211     Ronald Goedemondt: Ze bestaan echt (2008)  Ze...
54212     Make Your Own Bed (1944)  Walter and Vivian l...
54213     Nature's Fury: Storm of the Century (2006)  O...
Name: Movie, Length: 54214, dtype: object

In [ ]:
train_target

0         8
1        24
2         1
3         8
4         8
         ..
54209     5
54210    13
54211     7
54212     5
54213    12
Name: genre, Length: 54214, dtype: int64

In [ ]:
test_input=testDf['Movie']

In [ ]:
test_input

0         Edgar's Lunch (1998)  L.R. Brane loves his li...
1         La guerra de papá (1977)  Spain, March 1964: ...
2         Off the Beaten Track (2010)  One year in the ...
3         Meu Amigo Hindu (2015)  His father has died, ...
4         Er nu zhai (1955)  Before he was known intern...
                               ...                        
54195     "Tales of Light & Dark" (2013)  Covering mult...
54196     Der letzte Mohikaner (1965)  As Alice and Cor...
54197     Oliver Twink (2007)  A movie 169 years in the...
54198     Slipstream (1973)  Popular, but mysterious ro...
54199     Curitiba Zero Grau (2010)  Curitiba is a city...
Name: Movie, Length: 54200, dtype: object

Removing Stopwords

In [ ]:
import spacy
from spacy.lang.en.stop_words import STOP_WORDS

In [ ]:
len(STOP_WORDS)

326

In [ ]:
stop=spacy.load('en_core_web_sm')

In [ ]:
def preprocess(text):
  doc=stop(text)

  no_stop_words=[token.text for token in doc if not token.is_stop]
  return " ".join(no_stop_words)

In [ ]:
#trainDf['text']=trainDf.Movie.apply(preprocess)

In [ ]:
#testDf['text']=testDf.Movie.apply(preprocess)

In [ ]:
#train_input_new=trainDf['text']

In [ ]:
#test_input_new=testDf['text']

Converting Movie Coulum to Vector beacuse Machine Learning Models Doesn't understand text.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

In [ ]:
vector=CountVectorizer()

In [ ]:
train_input_count=vector.fit_transform(train_input)

In [ ]:
train_input_count.shape

(54214, 135743)

In [ ]:
test_input_count=vector.transform(test_input)

###Model Training :

####Logistic Regression :

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
model=LogisticRegression(solver='sag')

In [ ]:
model.fit(train_input_count,train_target)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


LogisticRegression(solver='sag')

In [ ]:
train_pred=model.predict(train_input_count)
train_pred

array([8, 8, 7, ..., 5, 5, 7])

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
accuracy_score(train_target,train_pred)

0.6762644335411517

In [ ]:
test_pred=model.predict(test_input_count)
test_pred

array([5, 8, 7, ..., 5, 8, 7])

Loading the Test Data Solution, In order to check the test accuracy because in our test data genre is not given.

In [ ]:
solDf=pd.read_table('test_data_solution.txt',delimiter=":::",names=['id','title','genre','description'])

<ipython-input-49-7ca950084569>:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  solDf=pd.read_table('test_data_solution.txt',delimiter=":::",names=['id','title','genre','description'])


In [ ]:
solDf.shape

(54200, 4)

In [ ]:
solDf.head()

id                          title          genre  \
0   1          Edgar's Lunch (1998)       thriller    
1   2      La guerra de papá (1977)         comedy    
2   3   Off the Beaten Track (2010)    documentary    
3   4        Meu Amigo Hindu (2015)          drama    
4   5             Er nu zhai (1955)          drama    

                                         description  
0   L.R. Brane loves his life - his car, his apar...  
1   Spain, March 1964: Quico is a very naughty ch...  
2   One year in the life of Albin and his family ...  
3   His father has died, he hasn't spoken with hi...  
4   Before he was known internationally as a mart...

In [ ]:
encoder = LabelEncoder()
solDf["genre"] = encoder.fit_transform(solDf["genre"])

In [ ]:
solDf.head()

id                          title  genre  \
0   1          Edgar's Lunch (1998)      24   
1   2      La guerra de papá (1977)       5   
2   3   Off the Beaten Track (2010)       7   
3   4        Meu Amigo Hindu (2015)       8   
4   5             Er nu zhai (1955)       8   

                                         description  
0   L.R. Brane loves his life - his car, his apar...  
1   Spain, March 1964: Quico is a very naughty ch...  
2   One year in the life of Albin and his family ...  
3   His father has died, he hasn't spoken with hi...  
4   Before he was known internationally as a mart...

In [ ]:
solDf.drop(columns=['id','title','description'],axis=1,inplace=True) #Dropping the useless ID

In [ ]:
test_target=solDf['genre'] #Creating the Test Target

In [ ]:
solDf.isna().sum()

genre    0
dtype: int64

In [ ]:
solDf.shape

(54200, 1)

In [ ]:
testDf.shape

(54200, 1)

In [ ]:
accuracy_score(test_target,test_pred)

0.5927859778597786

####Random Forest :

In [ ]:
#Random Forest
from sklearn.ensemble import RandomForestClassifier

In [ ]:
model=RandomForestClassifier(n_jobs=-1,random_state=42)

In [ ]:
%%time
model.fit(train_input_count,train_target)

CPU times: user 9min 38s, sys: 1.88 s, total: 9min 40s
Wall time: 5min 37s


RandomForestClassifier(n_jobs=-1, random_state=42)

In [ ]:
model.score(test_input_count,test_target)

0.4657380073800738

####KNeighbors Classifier :

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline

In [ ]:
model=KNeighborsClassifier()

In [ ]:
model.fit(train_input_count,train_target)

KNeighborsClassifier()

In [ ]:
train_pred=model.predict(train_input_count)

In [ ]:
accuracy_score(train_pred,train_target)

0.4978603312797432

In [ ]:
test_pred=model.predict(test_input_count)

In [ ]:
accuracy_score(test_pred,test_target)

0.32483394833948337

###Conclusion :

The following conclusion can be drawn :


*   The Data Set is Imbalanced.Hence, We are able to get only 59% Accuracy on Test Data.
*   The Dataset Has 13K data for Drama &  Documentary & Less than 200 for News and War.
*   Due Imbalance of Data The model is not reaching good accuracy.
*   The Removing of stopwords doesn't effect the accurcy more and on the other hand it takes lots of time. Therefore,I dropped removing Stopwords.


